File: app/src/main/java/com/tangem/presentation/activity/ConfirmPaymentActivity.java
package com.tangem.presentation.activity;

import android.app.Activity;
import android.content.Intent;
import android.nfc.NfcAdapter;
import android.nfc.Tag;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.Editable;
import android.text.Html;
import android.text.Spanned;
import android.text.TextWatcher;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.tangem.data.network.request.ElectrumRequest;
import com.tangem.data.network.request.FeeRequest;
import com.tangem.data.network.request.InfuraRequest;
import com.tangem.data.network.task.confirm_payment.ConnectFeeTask;
import com.tangem.data.network.task.confirm_payment.ConnectTask;
import com.tangem.data.network.task.confirm_payment.ETHRequestTask;
import com.tangem.domain.cardReader.NfcManager;
import com.tangem.domain.wallet.Blockchain;
import com.tangem.domain.wallet.CoinEngine;
import com.tangem.domain.wallet.CoinEngineFactory;
import com.tangem.domain.wallet.SharedData;
import com.tangem.domain.wallet.TangemCard;
import com.tangem.domain.wallet.Transaction;
import com.tangem.domain.wallet.UnspentOutputInfo;
import com.tangem.util.BTCUtils;
import com.tangem.util.DerEncodingUtil;
import com.tangem.util.FormatUtil;
import com.tangem.util.Util;
import com.tangem.wallet.R;

import java.io.IOException;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class ConfirmPaymentActivity extends AppCompatActivity implements NfcAdapter.ReaderCallback {

    private static final int REQUEST_CODE_SIGN_PAYMENT = 1;
    private static final int REQUEST_CODE_REQUEST_PIN2 = 2;
    public Button btnSend;
    public boolean feeRequestSuccess = false;
    public boolean balanceRequestSuccess = false;
    private EditText etWallet;
    //    private TextView tvAmountEquivalent;
    private TextView tvFeeEquivalent;
    public EditText etAmount;
    public EditText etFee;
    private ImageView ivCamera;
    public TangemCard mCard;
    public RadioGroup rgFee;
    public String minFee = null, maxFee = null, normalFee = null;
    public Long minFeeInInternalUnits = 0L;
    private NfcManager mNfcManager;
    private int requestPIN2Count = 0;
    public ProgressBar progressBar;
    public boolean nodeCheck = false;
    public Date dtVerifyed = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_confirm_payment);

        MainActivity.commonInit(getApplicationContext());
        mNfcManager = new NfcManager(this, this);

        mCard = new TangemCard(getIntent().getStringExtra("UID"));
        mCard.LoadFromBundle(getIntent().getExtras().getBundle("Card"));

        progressBar = findViewById(R.id.progressBar);
        btnSend = findViewById(R.id.btnSend);
        etWallet = findViewById(R.id.etWallet);
        TextView tvCardID = findViewById(R.id.tvCardID);
        TextView tvBalance = findViewById(R.id.tvBalance);
        TextView tvCurrency = findViewById(R.id.tvCurrency);
        TextView tvCurrency2 = findViewById(R.id.tvCurrency2);
        etAmount = findViewById(R.id.etAmount);
        etFee = findViewById(R.id.etFee);
//        TextView tvBalanceEquivalent = findViewById(R.id.tvBalanceEquivalent);
//        tvAmountEquivalent = findViewById(R.id.tvAmountEquivalent);
        tvFeeEquivalent = findViewById(R.id.tvFeeEquivalent);
        ivCamera = findViewById(R.id.ivCamera);
        rgFee = findViewById(R.id.rgFee);

        rgFee.setOnCheckedChangeListener((group, checkedId) -> doSetFee(checkedId));

        etAmount.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
//                try {
//
//                    CoinEngine engine = CoinEngineFactory.Create(mCard.getBlockchain());
//                    tvAmountEquivalent.setText(engine.GetAmountEqualentDescriptor(mCard, etAmount.getText().toString()));
//                    if (!mCard.getAmountEquivalentDescriptionAvailable()) {
//                        tvAmountEquivalent.setError("Service unavailable");
//                    } else {
//                        tvAmountEquivalent.setError(null);
//                    }
//                } catch (Exception e) {
//                    e.printStackTrace();
//                    tvAmountEquivalent.setText("");
//                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        etFee.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                try {

                    CoinEngine engine = CoinEngineFactory.Create(mCard.getBlockchain());
                    String eqFee = engine.EvaluteFeeEquivalent(mCard, etFee.getText().toString());
                    tvFeeEquivalent.setText(eqFee);

                    if (!mCard.getAmountEquivalentDescriptionAvailable()) {
                        tvFeeEquivalent.setError("Service unavailable");
                    } else {
                        tvFeeEquivalent.setError(null);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    tvFeeEquivalent.setText("");
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        CoinEngine engine = CoinEngineFactory.Create(mCard.getBlockchain());
        //tvBalance.setText(engine.GetBalanceWithAlter(mCard));
        if (mCard.getBlockchain() == Blockchain.Token) {
            Spanned html = Html.fromHtml(engine.GetBalanceWithAlter(mCard));
            tvBalance.setText(html);
        } else {
            tvBalance.setText(engine.GetBalanceWithAlter(mCard));
        }
        etAmount.setText(getIntent().getStringExtra(SignPaymentActivity.EXTRA_AMOUNT));
        tvCurrency.setText(engine.GetBalanceCurrency(mCard));
        tvCurrency2.setText(engine.GetFeeCurrency());

        tvCardID.setText(mCard.getCIDDescription());

        //tvBalanceEquivalent.setText(mCard.getBalanceEquivalentDescription());
//        tvBalanceEquivalent.setText(engine.GetBalanceEquivalent(mCard));
//        if (!mCard.getAmountEquivalentDescriptionAvailable()) {
//            tvBalanceEquivalent.setError("Service unavailable");
//        } else {
//            tvBalanceEquivalent.setError(null);
//        }

        etWallet.setText(getIntent().getStringExtra("Wallet"));

        etFee.setText("?");

        btnSend.setVisibility(View.INVISIBLE);
        feeRequestSuccess = false;
        balanceRequestSuccess = false;

        btnSend.setOnClickListener(v -> {
            Calendar calendar = Calendar.getInstance();
            calendar.add(Calendar.MINUTE, -1);

            if (dtVerifyed == null || dtVerifyed.before(calendar.getTime())) {
                finishActivityWithError(Activity.RESULT_CANCELED, "The obtained data is outdated! Try again");
                return;
            }

            CoinEngine engineCoin = CoinEngineFactory.Create(mCard.getBlockchain());

            if (engineCoin.IsNeedCheckNode() && !nodeCheck) {
                Toast.makeText(getBaseContext(), "Cannot reach current active blockchain node. Try again", Toast.LENGTH_LONG).show();
                return;
            }
            String txFee = etFee.getText().toString();
            String txAmount = etAmount.getText().toString();


            if (!engineCoin.HasBalanceInfo(mCard)) {
                finishActivityWithError(Activity.RESULT_CANCELED, "Cannot check balance! No connection with blockchain nodes");
                return;
            } else if (!engineCoin.IsBalanceNotZero(mCard)) {
                finishActivityWithError(Activity.RESULT_CANCELED, "The wallet is empty");
                return;
            } else if (!engineCoin.CheckUnspentTransaction(mCard)) {
                //else if (mCard.getUnspentTransactions().size() == 0 && mCard.getBlockchain() != Blockchain.Ethereum) {
                finishActivityWithError(Activity.RESULT_CANCELED, "Please wait for confirmation of incoming transaction");
                return;
            }

            if (!engineCoin.CheckAmountValie(mCard, txAmount, txFee, minFeeInInternalUnits)) {
                finishActivityWithError(Activity.RESULT_CANCELED, "Fee exceeds payment amount. Enter correct value and repeat sending.");
                return;
            }

            requestPIN2Count = 0;
            Intent intent = new Intent(getBaseContext(), RequestPINActivity.class);
            intent.putExtra("mode", RequestPINActivity.Mode.RequestPIN2.toString());
            intent.putExtra("UID", mCard.getUID());
            intent.putExtra("Card", mCard.getAsBundle());
            startActivityForResult(intent, REQUEST_CODE_REQUEST_PIN2);
        });

        if (mCard.getBlockchain() == Blockchain.Ethereum || mCard.getBlockchain() == Blockchain.EthereumTestNet || mCard.getBlockchain() == Blockchain.Token) {
            ETHRequestTask task = new ETHRequestTask(ConfirmPaymentActivity.this, mCard.getBlockchain());
            InfuraRequest req = InfuraRequest.GetGasPrise(mCard.getWallet());
            req.setID(67);
            req.setBlockchain(mCard.getBlockchain());
            rgFee.setEnabled(false);
            task.execute(req);

        } else {
            rgFee.setEnabled(true);

            SharedData data = new SharedData(SharedData.COUNT_REQUEST);

            CoinEngine engineCoin = CoinEngineFactory.Create(mCard.getBlockchain());

            for (int i = 0; i < data.allRequest; ++i) {

                String nodeAddress = engineCoin.GetNextNode(mCard);
                int nodePort = engineCoin.GetNextNodePort(mCard);
                ConnectTask connectTaskEx = new ConnectTask(ConfirmPaymentActivity.this, nodeAddress, nodePort, data);
                connectTaskEx.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, ElectrumRequest.CheckBalance(mCard.getWallet()));
            }

//            String nodeAddress = engineCoin.GetNode(mCard);
//            int nodePort = engineCoin.GetNodePort(mCard);
//            ConnectTask connectTask = new ConnectTask(nodeAddress, nodePort, data);
//
//
//            connectTask.execute(ElectrumRequest.GetFee(mCard.getWallet()));

            int calcSize = 256;
            try {
                calcSize = BuildSize(etWallet.getText().toString(), "0.00", etAmount.getText().toString());
            } catch (Exception ex) {
                Log.e("Build Fee error", ex.getMessage());
            }

            SharedData sharedFee = new SharedData(SharedData.COUNT_REQUEST);

            progressBar.setVisibility(View.VISIBLE);
            for (int i = 0; i < SharedData.COUNT_REQUEST; ++i) {

                ConnectFeeTask feeTask = new ConnectFeeTask(ConfirmPaymentActivity.this, sharedFee);

                feeTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR,
                        FeeRequest.GetFee(mCard.getWallet(), calcSize, FeeRequest.NORMAL),
                        FeeRequest.GetFee(mCard.getWallet(), calcSize, FeeRequest.MINIMAL),
                        FeeRequest.GetFee(mCard.getWallet(), calcSize, FeeRequest.PRIORITY));
            }
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mNfcManager.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        mNfcManager.onPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        mNfcManager.onStop();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_SIGN_PAYMENT) {
            if (data != null && data.getExtras() != null) {
                if (data.getExtras().containsKey("UID") && data.getExtras().containsKey("Card")) {
                    TangemCard updatedCard = new TangemCard(data.getStringExtra("UID"));
                    updatedCard.LoadFromBundle(data.getBundleExtra("Card"));
                    mCard = updatedCard;
                }
            }
            if (resultCode == SignPaymentActivity.RESULT_INVALID_PIN && requestPIN2Count < 2) {
                requestPIN2Count++;
                Intent intent = new Intent(getBaseContext(), RequestPINActivity.class);
                intent.putExtra("mode", RequestPINActivity.Mode.RequestPIN2.toString());
                intent.putExtra("UID", mCard.getUID());
                intent.putExtra("Card", mCard.getAsBundle());
                startActivityForResult(intent, REQUEST_CODE_REQUEST_PIN2);
                return;
            }
            setResult(resultCode, data);
            finish();
        } else if (requestCode == REQUEST_CODE_REQUEST_PIN2) {
            if (resultCode == Activity.RESULT_OK) {
                Intent intent = new Intent(getBaseContext(), SignPaymentActivity.class);
                intent.putExtra("UID", mCard.getUID());
                intent.putExtra("Card", mCard.getAsBundle());
                intent.putExtra("Wallet", etWallet.getText().toString());
                intent.putExtra(SignPaymentActivity.EXTRA_AMOUNT, etAmount.getText().toString());
                intent.putExtra("Fee", etFee.getText().toString());
                startActivityForResult(intent, REQUEST_CODE_SIGN_PAYMENT);
            } else {
                Toast.makeText(getBaseContext(), "PIN2 is required to sign the payment", Toast.LENGTH_LONG).show();
            }
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        switch (keyCode) {
            case KeyEvent.KEYCODE_BACK:
                Intent intent = new Intent();
//                intent.putExtra("message", "Operation canceled");
                setResult(Activity.RESULT_CANCELED, intent);
                finish();
                return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    public void finishActivityWithError(int errorCode, String message) {
        //Snackbar.make(etFee, message, Snackbar.LENGTH_LONG).show();
        Intent intent = new Intent();
        intent.putExtra("message", message);
        setResult(errorCode, intent);
        finish();
    }

    @Override
    public void onTagDiscovered(Tag tag) {
        try {
            Log.w(getClass().getName(), "Ignore discovered tag!");
            mNfcManager.ignoreTag(tag);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    int BuildSize(String outputAddress, String outFee, String outAmount) throws Exception {
        String myAddress = mCard.getWallet();
        String changeAddress = myAddress; //"n2eMqTT929pb1RDNuqEnxdaLau1rxy3efi";
        byte[] pbKey = mCard.getWalletPublicKey();
        byte[] pbComprKey = mCard.getWalletPublicKeyRar();

        // Build script for our address
        List<TangemCard.UnspentTransaction> rawTxList = mCard.getUnspentTransactions();
        byte[] outputScriptWeAreAbleToSpend = Transaction.Script.buildOutput(myAddress).bytes;

        // Collect unspent
        ArrayList<UnspentOutputInfo> unspentOutputs = BTCUtils.getOutputs(rawTxList, outputScriptWeAreAbleToSpend);

        long fullAmount = 0;
        for (int i = 0; i < unspentOutputs.size(); ++i) {
            fullAmount += unspentOutputs.get(i).value;
        }

        // Get first unspent
        UnspentOutputInfo outPut = unspentOutputs.get(0);
        int outPutIndex = outPut.outputIndex;

        // get prev TX id;
        String prevTXID = rawTxList.get(0).txID;//"f67b838d6e2c0c587f476f583843e93ff20368eaf96a798bdc25e01f53f8f5d2";

        long fees = FormatUtil.ConvertStringToLong(outFee);
        long amount = FormatUtil.ConvertStringToLong(outAmount);
        amount = amount - fees;

        long change = fullAmount - fees - amount;

        if (amount + fees > fullAmount) {
            throw new Exception(String.format("Balance (%d) < amount (%d) + (%d)", fullAmount, change, amount));
        }

        byte[][] hashesForSign = new byte[unspentOutputs.size()][];

        for (int i = 0; i < unspentOutputs.size(); ++i) {
            byte[] newTX = BTCUtils.buildTXForSign(myAddress, outputAddress, changeAddress, unspentOutputs, i, amount, change);

            byte[] hashData = Util.calculateSHA256(newTX);
            byte[] doubleHashData = Util.calculateSHA256(hashData);

            Log.e("TX_BODY_1", BTCUtils.toHex(newTX));
            Log.e("TX_HASH_1", BTCUtils.toHex(hashData));
            Log.e("TX_HASH_2", BTCUtils.toHex(doubleHashData));

            unspentOutputs.get(i).bodyDoubleHash = doubleHashData;
            unspentOutputs.get(i).bodyHash = hashData;

            hashesForSign[i] = doubleHashData;
        }

        byte[] signFromCard = new byte[64 * unspentOutputs.size()];

        for (int i = 0; i < unspentOutputs.size(); ++i) {
            BigInteger r = new BigInteger(1, Arrays.copyOfRange(signFromCard, 0 + i * 64, 32 + i * 64));
            BigInteger s = new BigInteger(1, Arrays.copyOfRange(signFromCard, 32 + i * 64, 64 + i * 64));
            byte[] encodingSign = DerEncodingUtil.packSignDer(r, s, pbKey);
            unspentOutputs.get(i).scriptForBuild = encodingSign;
        }

        byte[] realTX = BTCUtils.buildTXForSend(outputAddress, changeAddress, unspentOutputs, amount, change);

        return realTX.length;
    }

    public void doSetFee(int checkedRadioButtonId) {
        switch (checkedRadioButtonId) {
            case R.id.rbMinimalFee:
                if (minFee != null) etFee.setText(minFee);
                else etFee.setText("?");
                break;
            case R.id.rbNormalFee:
                if (normalFee != null) etFee.setText(normalFee);
                else etFee.setText("?");
                break;
            case R.id.rbMaximumFee:
                if (maxFee != null) etFee.setText(maxFee);
                else etFee.setText("?");
                break;
        }
    }

}

File: app/src/main/java/com/tangem/presentation/activity/ConfirmPaymentActivity.kt
package com.tangem.presentation.activity

import android.app.Activity
import android.content.Intent
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.Editable
import android.text.Html
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.widget.Toast
import com.tangem.Constant
import com.tangem.data.Blockchain
import com.tangem.domain.wallet.CoinEngine
import com.tangem.domain.wallet.CoinEngineFactory
import com.tangem.domain.wallet.TangemContext
import com.tangem.presentation.event.TransactionFinishWithError
import com.tangem.tangemcard.android.reader.NfcManager
import com.tangem.tangemcard.data.TangemCard
import com.tangem.tangemcard.data.loadFromBundle
import com.tangem.util.LOG
import com.tangem.util.UtilHelper
import com.tangem.wallet.R
import kotlinx.android.synthetic.main.activity_confirm_payment.*
import org.greenrobot.eventbus.EventBus
import java.io.IOException
import java.util.*

class ConfirmPaymentActivity : AppCompatActivity(), NfcAdapter.ReaderCallback {

    private var nfcManager: NfcManager? = null

    private lateinit var ctx: TangemContext
    private lateinit var amount: CoinEngine.Amount

    private var isIncludeFee: Boolean = true
    private var requestPIN2Count = 0
    private var nodeCheck = true
    private var dtVerified: Date? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_confirm_payment)

        nfcManager = NfcManager(this, this)

        ctx = TangemContext.loadFromBundle(this, intent.extras)

        val engine = CoinEngineFactory.create(ctx)

        val html = Html.fromHtml(engine!!.balanceHTML)
        tvBalance.text = html

        isIncludeFee = intent.getBooleanExtra(Constant.EXTRA_FEE_INCLUDED, true)

        if (isIncludeFee)
            tvIncFee.setText(R.string.including_fee)
        else
            tvIncFee.setText(R.string.not_including_fee)

        amount = CoinEngine.Amount(intent.getStringExtra(Constant.EXTRA_AMOUNT), intent.getStringExtra(Constant.EXTRA_AMOUNT_CURRENCY))

        if (ctx.blockchain == Blockchain.Token && amount.currency != "ETH")
            tvIncFee.visibility = View.INVISIBLE
        else
            tvIncFee.visibility = View.VISIBLE

        etAmount.setText(amount.toValueString())
        tvCurrency.text = engine.balanceCurrency
        tvCurrency2.text = engine.feeCurrency
        tvCardID.text = ctx.card!!.cidDescription
        etWallet.setText(intent.getStringExtra(Constant.EXTRA_TARGET_ADDRESS))
        etFee.setText("")

        btnSend.visibility = View.INVISIBLE

        rgFee.isEnabled = !(ctx.blockchain == Blockchain.Ethereum || ctx.blockchain == Blockchain.EthereumTestNet || ctx.blockchain == Blockchain.Token || ctx.blockchain == Blockchain.BitcoinCash || ctx.blockchain == Blockchain.Litecoin)

        // set listeners
        rgFee.setOnCheckedChangeListener { _, checkedId -> doSetFee(checkedId) }
        etFee.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                try {
                    val engine = CoinEngineFactory.create(ctx)
                    val eqFee = engine!!.evaluateFeeEquivalent(etFee!!.text.toString())
                    tvFeeEquivalent.text = eqFee

                    if (!ctx.coinData!!.amountEquivalentDescriptionAvailable) {
                        tvFeeEquivalent.error = "Service unavailable"
                        tvCurrency2.visibility = View.GONE
                        tvFeeEquivalent.visibility = View.GONE
                    } else
                        tvFeeEquivalent.error = null

                } catch (e: Exception) {
                    e.printStackTrace()
                    tvFeeEquivalent.text = ""
                }
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
        btnSend.setOnClickListener {
            val calendar = Calendar.getInstance()
            calendar.add(Calendar.MINUTE, -1)

            if (dtVerified == null || dtVerified!!.before(calendar.time)) {
                finishWithError(Activity.RESULT_CANCELED, getString(R.string.the_obtained_data_is_outdated_try_again))
                return@setOnClickListener
            }

            val engineCoin = CoinEngineFactory.create(ctx)

            if (engineCoin!!.isNeedCheckNode && !nodeCheck) {
                Toast.makeText(baseContext, getString(R.string.cannot_reach_current_active_blockchain_node_try_again), Toast.LENGTH_LONG).show()
                return@setOnClickListener
            }

            val txFee = engineCoin.convertToAmount(etFee.text.toString(), tvCurrency2.text.toString())
            val txAmount = engineCoin.convertToAmount(etAmount.text.toString(), tvCurrency.text.toString())

            if (!engineCoin.hasBalanceInfo()) {
                finishWithError(Activity.RESULT_CANCELED, getString(R.string.cannot_check_balance_no_connection_with_blockchain_nodes))
                return@setOnClickListener

            } else if (!engineCoin.isBalanceNotZero) {
                finishWithError(Activity.RESULT_CANCELED, getString(R.string.the_wallet_is_empty))
                return@setOnClickListener

            } else if (!engineCoin.isExtractPossible) {
                finishWithError(Activity.RESULT_CANCELED, getString(R.string.please_wait_for_confirmation_of_incoming_transaction))
                return@setOnClickListener
            }

            if (!engineCoin.checkNewTransactionAmountAndFee(txAmount, txFee, isIncludeFee)) {
                finishWithError(Activity.RESULT_CANCELED, getString(R.string.not_enough_funds_on_your_card))
                return@setOnClickListener
            }

            requestPIN2Count = 0
            val intent = Intent(baseContext, PinRequestActivity::class.java)
            intent.putExtra(Constant.EXTRA_MODE, PinRequestActivity.Mode.RequestPIN2.toString())
            ctx.saveToIntent(intent)
            intent.putExtra(Constant.EXTRA_FEE_INCLUDED, isIncludeFee)
            startActivityForResult(intent, Constant.REQUEST_CODE_REQUEST_PIN2_)
        }

        val coinEngine = CoinEngineFactory.create(ctx)

        progressBar.visibility = View.VISIBLE

        coinEngine!!.requestFee(
                object : CoinEngine.BlockchainRequestsCallbacks {
                    override fun onComplete(success: Boolean) {
                        if (success) {
                            onProgress()
                            progressBar.visibility = View.INVISIBLE
                            dtVerified = Date()
                        } else {
                            finishWithError(Activity.RESULT_CANCELED, ctx.error)
                        }
                    }

                    override fun onProgress() {
                        doSetFee(rgFee.checkedRadioButtonId)
                    }

                    override fun allowAdvance(): Boolean {
                        return UtilHelper.isOnline(this@ConfirmPaymentActivity)
                    }
                },
                etWallet.text.toString(),
                amount)
    }

    public override fun onResume() {
        super.onResume()
        nfcManager!!.onResume()
    }

    public override fun onPause() {
        super.onPause()
        nfcManager!!.onPause()
    }

    public override fun onStop() {
        super.onStop()
        nfcManager!!.onStop()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == Constant.REQUEST_CODE_SIGN_PAYMENT) {
            if (data != null && data.extras != null) {
                if (data.extras!!.containsKey("UID") && data.extras!!.containsKey("Card")) {
                    val updatedCard = TangemCard(data.getStringExtra("UID"))
                    updatedCard.loadFromBundle(data.getBundleExtra("Card"))
                    ctx.card = updatedCard
                }
            }
            if (resultCode == Constant.RESULT_INVALID_PIN_ && requestPIN2Count < 2) {
                requestPIN2Count++

                val intent = Intent(baseContext, PinRequestActivity::class.java)
                intent.putExtra(Constant.EXTRA_MODE, PinRequestActivity.Mode.RequestPIN2.toString())
                ctx.saveToIntent(intent)
                intent.putExtra(Constant.EXTRA_FEE_INCLUDED, isIncludeFee)
                startActivityForResult(intent, Constant.REQUEST_CODE_REQUEST_PIN2_)

                return
            }
            setResult(resultCode, data)
            finish()
        } else if (requestCode == Constant.REQUEST_CODE_REQUEST_PIN2_) {
            if (resultCode == Activity.RESULT_OK) {
                val intent = Intent(baseContext, SignPaymentActivity::class.java)
                ctx.saveToIntent(intent)
                intent.putExtra(Constant.EXTRA_TARGET_ADDRESS, etWallet!!.text.toString())
                intent.putExtra(Constant.EXTRA_AMOUNT, etAmount.text.toString())
                intent.putExtra(Constant.EXTRA_AMOUNT_CURRENCY, tvCurrency.text.toString())
                intent.putExtra(Constant.EXTRA_FEE, etFee.text.toString())
                intent.putExtra(Constant.EXTRA_FEE_CURRENCY, tvCurrency2.text.toString())
                intent.putExtra(Constant.EXTRA_FEE_INCLUDED, isIncludeFee)
                startActivityForResult(intent, Constant.REQUEST_CODE_SIGN_PAYMENT)
            } else
                Toast.makeText(baseContext, R.string.pin_2_is_required_to_sign_the_payment, Toast.LENGTH_LONG).show()
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_BACK -> {
                val intent = Intent()
                setResult(Activity.RESULT_CANCELED, intent)
                finish()
                return true
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onTagDiscovered(tag: Tag) {
        try {
            nfcManager!!.ignoreTag(tag)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun doSetFee(checkedRadioButtonId: Int) {
        var txtFee = ""
        when (checkedRadioButtonId) {
            R.id.rbMinimalFee ->
                if (ctx.coinData.minFee != null) {
                    txtFee = ctx.coinData.minFee!!.toValueString()
                    btnSend.visibility = View.VISIBLE
                } else
                    btnSend.visibility = View.INVISIBLE

            R.id.rbNormalFee ->
                if (ctx.coinData.normalFee != null) {
                    txtFee = ctx.coinData.normalFee!!.toValueString()
                    btnSend.visibility = View.VISIBLE
                } else
                    btnSend.visibility = View.INVISIBLE

            R.id.rbMaximumFee ->
                if (ctx.coinData.maxFee != null) {
                    txtFee = ctx.coinData.maxFee!!.toValueString()
                    btnSend.visibility = View.VISIBLE
                } else
                    btnSend.visibility = View.INVISIBLE
        }
        etFee.setText(txtFee.replace(',', '.'))
    }

    private fun finishWithError(errorCode: Int, message: String) {
        val transactionFinishWithError = TransactionFinishWithError()
        transactionFinishWithError.message = message
        EventBus.getDefault().post(transactionFinishWithError)

        val intent = Intent()
        intent.putExtra("message", message)
        setResult(errorCode, intent)
        finish()
    }

}